types.go

```diff
@@ -0,0 +1,476 @@
+		r.registerType(TypeText, "text", SimpleCQLType{varcharLikeCQLType{
+			typ: TypeText,
+		}})
+		r.registerType(TypeTime, "time", SimpleCQLType{timeCQLType{}})
+		r.registerType(TypeTimestamp, "timestamp", SimpleCQLType{timestampCQLType{}})
+		r.registerType(TypeTimeUUID, "timeuuid", SimpleCQLType{timeUUIDCQLType{}})
+		r.registerType(TypeTinyInt, "tinyint", SimpleCQLType{tinyIntCQLType{}})
+		r.registerType(TypeUUID, "uuid", SimpleCQLType{uuidCQLType{}})
+		r.registerType(TypeVarchar, "varchar", SimpleCQLType{varcharLikeCQLType{
+			typ: TypeVarchar,
+		}})
+		r.registerType(TypeVarint, "varint", SimpleCQLType{varintCQLType{}})
+		r.registerType(TypeUDT, "udt", udtCQLType{})
+
+		// these types need references to the registered types
+		r.registerType(TypeList, "list", listSetCQLType{
+			typ:   TypeList,
+			types: r,
+		})
+		r.registerType(TypeSet, "set", listSetCQLType{
+			typ:   TypeSet,
+			types: r,
+		})
+		r.registerType(TypeMap, "map", mapCQLType{
+			types: r,
+		})
+		r.registerType(TypeTuple, "tuple", tupleCQLType{
+			types: r,
+		})
+
+		// these are all for Cassandra 2.x and older
+		r.byString["AsciiType"] = TypeAscii
+		r.byString["LongType"] = TypeBigInt
+		r.byString["BytesType"] = TypeBlob
+		r.byString["BooleanType"] = TypeBoolean
+		r.byString["CounterColumnType"] = TypeCounter
+		r.byString["DecimalType"] = TypeDecimal
+		r.byString["DoubleType"] = TypeDouble
+		r.byString["FloatType"] = TypeFloat
+		r.byString["InetAddressType"] = TypeInet
+		r.byString["Int32Type"] = TypeInt
+		r.byString["ShortType"] = TypeSmallInt
+		r.byString["UTF8Type"] = TypeVarchar
+		r.byString["TimeType"] = TypeTime
+		r.byString["TimestampType"] = TypeTimestamp
+		// DateType was a timestamp when date didn't exist
+		r.byString["DateType"] = TypeTimestamp
+		r.byString["TimeUUIDType"] = TypeTimeUUID
+		r.byString["ByteType"] = TypeTinyInt
+		r.byString["UUIDType"] = TypeUUID
+		r.byString["LexicalUUIDType"] = TypeUUID
+		r.byString["IntegerType"] = TypeVarint
+		r.byString["UserType"] = TypeUDT
+		r.byString["ListType"] = TypeList
+		r.byString["MapType"] = TypeMap
+		r.byString["SetType"] = TypeSet
+		r.byString["TupleType"] = TypeTuple
+	})
+}
+
+// RegisterType registers a new CQL data type. Type should be the CQL id for
+// the type. Name is the name of the type as returned in the metadata for the
+// column. CQLType is the implementation of the type.
+// This function must not be called after a session has been created.
+func (r *RegisteredTypes) RegisterType(typ Type, name string, t CQLType) {
+	r.init()
+	r.mut.Lock()
+	defer r.mut.Unlock()
+	r.registerType(typ, name, t)
+}
+
+func (r *RegisteredTypes) registerType(typ Type, name string, t CQLType) {
+	if typ == TypeCustom {
+		panic("custom types must be registered with RegisterCustom")
+	}
+
+	if _, ok := r.byType[typ]; ok {
+		panic(fmt.Errorf("type %d already registered", typ))
+	}
+	if _, ok := r.byString[name]; ok {
+		panic(fmt.Errorf("type name %s already registered", name))
```

Comment:
should probably return errors instead of panic

Reply:
I'm in favor of panic only because these should happen at init and there's nothing else to do but panic and the stack trace will probably be helpful for anyone calling this.

Reply:
Done!